### PR TITLE
Spark 3.4: Initial support

### LINKED ISSUE #7174: Add support for Spark 3.4
### Feature Request / Improvement

We should add support for Spark 3.4.

### Query engine

Spark

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds initial support for Spark 3.4 and consists of 3 commits that must be preserved while merging.

- a0012e67ddf9f7e2ff1061938a126e45259c106c moves 3.3 classes to 3.4 folder
- c6aa3f1a1bf3d266a03107dc2cb39403d54ecbb4 copies 3.4 folder back into 3.3
- 723487b2452fb8b2b6ca04ee71d63b9236531307 makes 3.4 work

The last change is the most important to review.

Note that this approach preserves the commit history only for 3.4 (our new default version). There are [tricks](https://stackoverflow.com/questions/16937359/git-copy-file-preserving-history) to keep history both for 3.3 and 3.4 but they may cause issues while rebasing. That's why I followed the exact [approach](https://github.com/apache/iceberg/pull/5094) we used in 3.3.

It is worth mentioning that the DROP table behavior in the Spark session catalog is broken. That's why some tests had to be adapted. We are exploring a Spark fix at the moment.

Resolves #7174.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @nastra @rdblue @singhpk234 @RussellSpitzer @flyrain @amogh-jahagirdar @jackye1995 @szehon-ho

**Review by RussellSpitzer — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Note to reviewers, Russell left comments directly on https://github.com/apache/iceberg/commit/723487b2452fb8b2b6ca04ee71d63b9236531307. I'll create a JIRA for the Spark issue so that we can reference it in currently ignored tests.

**Review by Fokko — APPROVED** [review]
Thanks for adding this @aokolnychyi! Should we do a performance benchmark at some point between 3.3 and 3.4?

**Comment by bryanck** [comment]
The TPC-DS benchmarking tool we currently use is EMR-specific unfortunately.

**Comment by aokolnychyi** [comment]
I merged this PR as it is really hard to keep up with changes in master. I had started from scratch multiple times. We will need to cherry-pick #6480 back to 3.3 as it is only in 3.4 after this PR. I'll do that next, it is easier than doing the copy again.

**Review by singhpk234 — COMMENTED** [review]
LGTM as well, Thanks a ton for adding this @aokolnychyi !

Just some minor comments on [723487b](https://github.com/apache/iceberg/commit/723487b2452fb8b2b6ca04ee71d63b9236531307) and 3.4 directory

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/data/parquet/SparkParquetWritersFlatDataBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+/**
+ * A benchmark that evaluates the performance of writing Parquet data with a flat schema using
+ * Iceberg and Spark Parquet writers.
+ *
+ * <p>To run this benchmark for spark-3.3: <code>
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/data/parquet/SparkParquetWritersFlatDataBenchmark.java:54]: Good catch! I'll follow up.

**Comment by aokolnychyi** [comment]
@singhpk234, I replied to comments on the change. Let me know if that makes sense, I'll follow up to fix JMH Javadoc.

**Comment by singhpk234** [comment]
> @singhpk234, I replied to comments on the change. Let me know if that makes sense

Makes sense to me. Thanks @aokolnychyi !

**Comment by mgorsk1** [comment]
thanks for this @aokolnychyi any eta on when we can expect iceberg-spark-runtime-3.14_2.12:1.2.0 in maven?

**Comment by vakarisbk** [comment]
I'm also interested in spark-runtime for 3.4.0

**Comment by aokolnychyi** [comment]
@mgorsk1 @vakarisbk, the plan to get a public release out mid May.

**Pull request merged**